kombu/transport/qpid.py

```diff
@@ -1205,66 +1210,42 @@ def __init__(self, **connection_options):
-
-        for sasl_mech in sasl_mechanisms:
-            try:
-                logger.debug(
-                    'Attempting to connect to qpid '
-                    'with SASL mechanism %s', sasl_mech,
-                )
-                modified_conn_opts = self.connection_options
-                modified_conn_opts['sasl_mechanisms'] = sasl_mech
-                self._qpid_conn = establish(**modified_conn_opts)
-                # connection was successful if we got this far
-                logger.info(
-                    'Connected to qpid with SASL mechanism %s', sasl_mech,
-                )
-                break
-            except ConnectionError as conn_exc:
-                # if we get one of these errors, do not raise an exception.
-                # Raising will cause the connection to be retried. Instead,
-                # just continue on to the next mech.
-                coded_as_auth_failure = getattr(conn_exc, 'code', None) == 320
-                contains_auth_fail_text = \
-                    'Authentication failed' in conn_exc.text
-                contains_mech_fail_text = \
-                    'sasl negotiation failed: no mechanism agreed' \
-                    in conn_exc.text
-                contains_mech_unavail_text = 'no mechanism available' \
-                    in conn_exc.text
-                if coded_as_auth_failure or \
-                        contains_auth_fail_text or contains_mech_fail_text or \
-                        contains_mech_unavail_text:
-                    logger.debug(
-                        'Unable to connect to qpid with SASL mechanism %s',
-                        sasl_mech,
-                    )
-                    continue
-                raise
-
-        if not self.get_qpid_connection():
-            logger.error(
-                'Unable to authenticate to qpid using '
-                'the following mechanisms: %s', sasl_mechanisms,
-            )
-            raise AuthenticationFailure(sys.exc_info()[1])
+        # There are several inconsistent behaviors in the sasl libraries
+        # used on different systems. This implementation uses only
+        # advertises one type to the server either ANONYMOUS or PLAIN.
+
+        sasl_mech = connection_options['sasl_mechanisms']
+
+        try:
+            msg = _('Attempting to connect to qpid with '
+                    'SASL mechanism %s') % sasl_mech
+            logger.debug(msg)
+            self._qpid_conn = establish(**self.connection_options)
+            # connection was successful if we got this far
+            msg = _('Connected to qpid with SASL '
+                    'mechanism %s') % sasl_mech
+            logger.info(msg)
+        except ConnectionError as conn_exc:
+            # if we get one of these errors, do not raise an exception.
+            # Raising will cause the connection to be retried. Instead,
+            # just continue on to the next mech.
```

Comment:
That is correct we only do one now based on how the broker string is configured. Letting the client libraries auto-negotiate has been hazardous because they all behave a little differently.